src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,353 @@
+      @Nullable final Character quote,
+      @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this.source = (AmazonS3Source) Preconditions.checkNotNull(source, "source");
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+
+    this.delimiter = MoreObjects.firstNonNull(delimiter, CSVFormat.DEFAULT.getDelimiter());
+    this.quote = MoreObjects.firstNonNull(quote, CSVFormat.DEFAULT.getQuoteCharacter());
+    this.escape = escape != null ? escape : CSVFormat.DEFAULT.getEscapeCharacter();
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+
+    this.partitionStartByteRange = partitionStartByteRange;
+    this.partitionEndByteRange = partitionEndByteRange;
+    this.isLastWorker = isLastWorker;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws IOException, DbException {
+    long lineNumberBegin = lineNumber;
+    boolean fixingStartByte = false;
+    boolean onLastRow = false;
+
+    boolean discardedRecord = false;
+
+    while ((buffer.numTuples() < TupleBatch.BATCH_SIZE)) {
+      discardedRecord = false;
+      lineNumber++;
+      if (parser.isClosed()) {
+        break;
+      }
+      try {
+        if (!iterator.hasNext()) {
+          parser.close();
+          break;
+        }
+      } catch (final RuntimeException e) {
+        throw new DbException("Error parsing row " + lineNumber, e);
+      }
+      CSVRecord record = iterator.next();
+      // This covers the case where the first row of a worker matches the schema. We only want to read this row if the
+      // previous character is '\n' or '\r'
+      if (record.size() == schema.numColumns()
+          && lineNumber - 1 == 0
+          && partitionStartByteRange != 0) {
+        InputStreamReader startStreamReader =
+            new InputStreamReader(
+                source.getInputStream(partitionStartByteRange - 1, partitionEndByteRange));
+        char currentChar = (char) startStreamReader.read();
+        if (currentChar != '\n' && currentChar != '\r') {
+          discardedRecord = true;
+        }
+      }
+      // This covers the case where the last row matches the schema's number of columns. We need to ensure that we read
+      // the entire row completely
+      else if (record.size() == schema.numColumns() && !iterator.hasNext() && !isLastWorker) {
+        long movingEndByte = partitionEndByteRange;
+        long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+        boolean newLineFound = false;
+        while (!newLineFound) {
+          movingEndByte += MyriaConstants.BYTE_OVERLAP_PARALLEL_INGEST;
+          // Create a stream to look for the new line
+          InputStream trailingEndInputStream =
+              source.getInputStream(partitionEndByteRange, movingEndByte);
+          InputStreamReader startStreamReader = new InputStreamReader(trailingEndInputStream);
+          int dataChar = startStreamReader.read();
```

Comment:
Again, `read()` could fail here if we start in the middle of a multibyte sequence, or it could continue into the current range (if the preceding byte is the beginning of a multibyte sequence). Ideally tests would cover these scenarios.